// @gemini-wallet/core/src/index.ts
// Main exports
export { Communicator } from "./communicator";

// Provider exports
export { GeminiWalletProvider } from "./provider";
export * from "./provider/provider.utils";

// Wallet exports
export { GeminiWallet } from "./wallets";

// Storage exports
export type { IStorage } from "./storage";
export {
  GeminiStorage,
  STORAGE_ETH_ACCOUNTS_KEY,
  STORAGE_ETH_ACTIVE_CHAIN_KEY,
  STORAGE_PASSKEY_CREDENTIAL_KEY,
  STORAGE_SETTINGS_KEY,
  STORAGE_SMART_ACCOUNT_KEY,
} from "./storage";

// Type exports
export type {
  AppContext,
  AppMetadata,
  Chain,
  ConnectResponse,
  GeminiProviderConfig,
  GeminiSdkAppContextMessage,
  GeminiSdkMessage,
  GeminiSdkMessageResponse,
  GeminiSdkSendTransaction,
  GeminiSdkSignMessage,
  GeminiSdkSignTypedData,
  GeminiSdkSwitchChain,
  PlatformType,
  ProviderEventCallback,
  ProviderEventMap,
  ProviderInterface,
  ProviderRpcError,
  ReverseEnsResponse,
  RpcRequestArgs,
  SendTransactionResponse,
  SignMessageResponse,
  SignTypedDataResponse,
  SwitchChainResponse,
} from "./types";
export { GeminiSdkEvent, ProviderEventEmitter } from "./types";

// Utility exports
export type {
  CalculateWalletAddressParams,
  WebAuthnValidatorData,
} from "./utils";
export {
  base64ToHex,
  bufferToBase64URLString,
  calculateWalletAddress,
  closePopup,
  decodeBase64,
  encodeBase64,
  generateAuthenticatorIdHash,
  generateRequestId,
  hexStringFromNumber,
  openPopup,
  reverseResolveEns,
  safeJsonStringify,
  utf8StringToBuffer,
  validateWebAuthnKey,
} from "./utils";

// Constants
export {
  DEFAULT_CHAIN_ID,
  POPUP_HEIGHT,
  POPUP_WIDTH,
  SDK_BACKEND_URL,
  SDK_VERSION,
} from "./constants";
